pkg/api/service.go

```diff
@@ -72,6 +73,78 @@ func (s *Service) Close() {
 	s.log.Info("closed")
 }
 
+// Pulls from DB and sends to client, updating the query's last seen cursor, until
+// the stream has caught up to the latest in the database.
+func (s *Service) catchUpFromCursor(
+	stream message_api.ReplicationApi_SubscribeEnvelopesServer,
+	query *message_api.EnvelopesQuery,
+	logger *zap.Logger,
+) error {
+	cursor := query.GetLastSeen().GetNodeIdToSequenceId()
+	if cursor == nil {
+		// Requester only wants new envelopes
+		return nil
+	}
+
+	for {
+		rows, err := s.fetchEnvelopes(stream.Context(), query, maxRequestedRows)
+		if err != nil {
+			return err
+		}
+		envs := make([]*envelopes.OriginatorEnvelope, 0, len(rows))
+		for _, r := range rows {
+			env, err := envelopes.NewOriginatorEnvelopeFromBytes(r.OriginatorEnvelope)
+			if err != nil {
+				// We expect to have already validated the envelope when it was inserted
+				logger.Error("could not unmarshal originator envelope", zap.Error(err))
+				continue
+			}
+			envs = append(envs, env)
+		}
+		err = s.sendEnvelopes(stream, query, envs)
+		if err != nil {
+			return status.Errorf(codes.Internal, "error sending envelopes: %v", err)
+		}
+		if len(rows) < int(maxRequestedRows) {
+			// There were no more envelopes in DB at time of fetch
+			break
+		}
+		// TODO(rich): Determine default rate limits and set query interval to match
+		time.Sleep(50 * time.Millisecond)
+	}
+	return nil
+}
+
+func (s *Service) sendEnvelopes(
+	stream message_api.ReplicationApi_SubscribeEnvelopesServer,
+	query *message_api.EnvelopesQuery,
+	envs []*envelopes.OriginatorEnvelope,
+) error {
+	cursor := query.GetLastSeen().GetNodeIdToSequenceId()
+	if cursor == nil {
+		cursor = make(map[uint32]uint64)
+		query.LastSeen = &message_api.VectorClock{
+			NodeIdToSequenceId: cursor,
+		}
+	}
+	for _, env := range envs {
+		if cursor[uint32(env.OriginatorNodeID())] >= env.OriginatorSequenceID() {
+			continue
+		}
+
+		// TODO(rich): Either batch send envelopes, or modify stream proto to
+		// send one envelope at a time.
+		err := stream.Send(&message_api.SubscribeEnvelopesResponse{
+			Envelopes: []*message_api.OriginatorEnvelope{env.Proto()},
+		})
```

Comment:
I can restructure this loop to batch send, I'm just unsure if that is more performant or not. We would be trading off the time to copy the envelopes to another array, versus the efficiency gain of batch sending. I'm guessing the efficiency gain will be greater, but I'm really unsure